tests/integration/taint_test.go

```diff
@@ -0,0 +1,384 @@
+func TestUntaintResource(t *testing.T) {
+	t.Parallel()
+
+	dir := t.TempDir()
+
+	// Write a simple Pulumi program
+	program := `
+const pulumi = require("@pulumi/pulumi");
+
+class MyResource extends pulumi.CustomResource {
+    constructor(name, args, opts) {
+        super("test:index:MyResource", name, args, opts);
+    }
+}
+
+const resource1 = new MyResource("myresource", {
+    prop1: "value1",
+});
+
+exports.resource1Id = resource1.id;
+`
+
+	err := os.WriteFile(filepath.Join(dir, "index.js"), []byte(program), 0600)
+	require.NoError(t, err)
+
+	// Write package.json
+	packageJSON := `{
+  "name": "untaint-test",
+  "main": "index.js",
+  "dependencies": {
+    "@pulumi/pulumi": "^3.0.0"
+  }
+}
+`
+	err = os.WriteFile(filepath.Join(dir, "package.json"), []byte(packageJSON), 0600)
+	require.NoError(t, err)
+
+	// Write Pulumi.yaml
+	pulumiYAML := `name: untaint-test
+runtime: nodejs
+description: Test untaint functionality
+`
+	err = os.WriteFile(filepath.Join(dir, "Pulumi.yaml"), []byte(pulumiYAML), 0600)
+	require.NoError(t, err)
+
+	e := &integration.ProgramTestOptions{
+		Dir:             dir,
+		Dependencies:    []string{"@pulumi/pulumi"},
+		Quick:           true,
+		SkipRefresh:     true,
+		NoParallel:      true,
+
+		ExtraRuntimeValidation: func(t *testing.T, stack integration.RuntimeValidationStackInfo) {
+			// Get the URN of the resource
+			var resourceURN string
+			for _, res := range stack.Deployment.Resources {
+				if res.Type == "test:index:MyResource" {
+					resourceURN = string(res.URN)
+					break
+				}
+			}
+			require.NotEmpty(t, resourceURN, "MyResource URN not found")
+
+			// Taint the resource
+			stdout, stderr, err := stack.RunCommand("pulumi", "state", "taint", resourceURN, "--yes")
+			require.NoError(t, err, "Failed to taint resource: %s\n%s", stdout, stderr)
+
+			// Preview should show it will be replaced
+			stdout, stderr, err = stack.RunCommand("pulumi", "preview", "--json")
+			require.NoError(t, err, "Failed to preview: %s\n%s", stdout, stderr)
+			assert.Contains(t, stdout, "replace", "Preview should show resource will be replaced")
+
+			// Untaint the resource
+			stdout, stderr, err = stack.RunCommand("pulumi", "state", "untaint", resourceURN, "--yes")
+			require.NoError(t, err, "Failed to untaint resource: %s\n%s", stdout, stderr)
+			assert.Contains(t, stdout, "1 resources untainted", "Expected success message")
+
+			// Preview should now show no changes
+			stdout, stderr, err = stack.RunCommand("pulumi", "preview")
+			require.NoError(t, err, "Failed to preview after untaint: %s\n%s", stdout, stderr)
+			assert.Contains(t, strings.ToLower(stdout), "no changes", "Should have no changes after untainting resource")
+		},
+	}
+
+	integration.ProgramTest(t, e)
+}
+
+// TestUntaintAll tests the --all flag for untaint
+func TestUntaintAll(t *testing.T) {
+	t.Parallel()
+
+	dir := t.TempDir()
+
+	// Write a program with multiple resources
+	program := `
```

Comment:
Usually we put this code in a new folder under `testdata`.  Any reason not to do so here?

Reply:
What is the benefit of putting it in testdata, versus inline? Seems this way there are fewer moving pieces?

Reply:
Mostly consistency with the other tests, but also it's easier to test things manually.  Not super important though.